src/include/enclave/ticketlock.h

```diff
@@ -1,55 +1,111 @@
+ * numbered ticket and wait until that number is called.  Threads that attempt
+ * to acquire the lock 'take a ticket' by atomically incrementing one counter
+ * and then wait until that ticket is the current one.  Threads releasing the
+ * lock increment a second counter representing the current ticket.
+ */
 struct ticketlock
 {
+    /**
+     * Anonymous union of the two words used to represent the ticket lock.
+     * It's not clear whether doing different-sized atomic accesses to the same
+     * address is defined behaviour in C.  It is definitely broken on some
+     * PowerPC implementations but is fine on x86.
+     */
     union {
-        uint64_t u;
+        /**
+         * A single view of the ticket and users fields.  This is needed for
+         * `ticket_trylock`, which attempts to atomically test `ticket` and set
+         * `users`.
+         */
+        _Atomic(uint64_t) u;
         struct
         {
-            uint32_t ticket;
-            uint32_t users;
+            /**
+             * The current ticket value.  Each thread that attempts to acquire
+             * a ticket lock waits until this field matches its ticket.
+             */
+            _Atomic(uint32_t) ticket;
+            /**
+             * The value that the next thread attempting to acquire the lock
+             * should take.  A thread 'takes a ticket' by atomically
+             * incrementing this field.  The value before the increment is the
+             * value of that thread's ticket.
+             */
+            _Atomic(uint32_t) users;
         } s;
     };
 #if DEBUG
     struct lthread* lt; // Thread that is holding the lock.
 #endif                  /* DEBUG */
 };
 
+/**
+ * Acquire a ticket lock.  This spins until the lock is acquired, it does *not*
+ * yield to other cooperatively scheduled threads.
+ */
 static inline void ticket_lock(struct ticketlock* t)
 {
-    uint32_t me = a_fetch_add_uint(&t->s.users, 1);
+    // Acquire a ticket
+    uint32_t me = atomic_fetch_add(&t->s.users, 1);
+    // Wait until our ticket is called.
     while (t->s.ticket != me)
-        a_spin();
+    {
+#ifdef __x86__
+        __builtin_ia32_pause()
+#endif
+            ;
+    }
+
 #if DEBUG
     t->lt = NULL;
     // t->lt = lthread_self();
 #endif /* DEBUG */
 }
 
+/**
+ * Release the ticket lock.
+ */
 static inline void ticket_unlock(struct ticketlock* t)
 {
-    a_barrier();
     t->s.ticket++;
 #if DEBUG
     t->lt = NULL;
 #endif /* DEBUG */
 }
 
+/**
+ * Try to acquire a ticket lock.  Returns zero on success, non-zero on failure.
+ */
 static inline int ticket_trylock(struct ticketlock* t)
 {
+    // Get the expected next ticket value
     uint32_t me = t->s.users;
+    // Set our exected next-ticket value.
     uint32_t menew = me + 1;
-    uint64_t cmp = ((uint64_t)me << 32) + me;
-    uint64_t cmpnew = ((uint64_t)menew << 32) + me;
+    // If the ticket lock is not held, t->s.users will equal t->s.ticket.
```

Comment:
This is relying on zero-initialized memory for the initial state of the `ticketlock` structure. Should we call that out?

Reply:
Also t->s.users == t->s.ticket is what it means to be in the locked state if you are using `ticket_lock`. Are these two ways of acquiring locks supposed to be used interchangeably for the same lock?

Reply:
Pls ignore my second comment, I misunderstood what `atomic_fetch_add` does.

Reply:
It doesn't actually require zero init, it requires that the two fields be initialised to the same value.  Zero init is fine though.